src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -650,6 +661,116 @@ vertexSet = Set.fromAscList . vertexList
+                                 Nothing -> dfs g p v
+                                 Just q  -> if q /= p
+                                            then return (u, v)
+                                            else mzero
+
+        dfs :: Ord a => AM.AdjacencyMap a -> Part -> a -> PartMonad a (a, a)
+        dfs g p v = do modify $ Map.insert v p
+                       let q = otherPart p
+                       asum $ map (action g q v) $ neighbours v g
+
+        runDfs :: Ord a => AM.AdjacencyMap a -> a -> PartMonad a (a, a, a)
+        runDfs g v = mod v <$> do m <- get
+                                  guard $ v `Map.notMember` m
+                                  dfs g LeftPart v
+
+        mod :: a -> (b, c) -> (a, b, c)
+        mod x (y, z) = (x, y, z)
+
+type PathMonad a = MaybeT (State (Set.Set a)) [a]
+
+findPath :: Ord a => a -> a -> AM.AdjacencyMap a -> Maybe [a]
+findPath u v g = evalState (runMaybeT $ dfs g v u) Set.empty
+    where
+        dfs :: Ord a => AM.AdjacencyMap a -> a -> a -> PathMonad a
+        dfs g t v | v == t = return [v]
+                  | otherwise = ((:) v) <$> do vis <- get
+                                               guard $ v `Set.notMember` vis
+                                               modify $ Set.insert v
+                                               asum $ map (dfs g t) $ neighbours v g
+
+-- | Test bipartiteness of given graph. In case of success, return an
+-- 'AdjacencyMap' with the same set of edges and each vertex marked with the
+-- part it belongs to. In case of failure, return any odd cycle in the graph.
+--
+-- The returned partition is lexicographicaly minimal. That is, consider the
+-- string of part identifiers for each vertex in ascending order. Then,
+-- considering that the identifier of the left part is less then the identifier
+-- of the right part, this string is lexicographically minimal of all such
+-- strings for all partitions.
+--
+-- /Note/: as 'AdjacencyMap' only represents __undirected__ bipartite graphs,
+-- all edges in the input graph are assumed to be bidirected and all edges in
+-- the output 'AdjacencyMap' are bidirected.
+--
+-- It is advised to use 'leftVertexList' and 'rightVertexList' to obtain the
+-- partition of the vertices and 'hasLeftVertex' and 'hasRightVertex' to check
+-- whether a vertex belongs to a part.
+--
+-- Complexity: /O((n + m) log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- detectParts 'Algebra.Graph.AdjacencyMap.empty'                                      == Right 'empty'
+-- detectParts ('Algebra.Graph.AdjacencyMap.vertex' x)                                 == Right ('leftVertex' x)
+-- detectParts (1 * (2 + 3))                              == Right ('edges' [(1, 2), (1, 3)])
+-- detectParts ((1 + 3) * (2 + 4) + 6 * 5)                == Right ('swap' (1 + 3) * (2 + 4) + 5 * 6)
+-- 'Data.Either.isRight' (detectParts ('Algebra.Graph.AdjacencyMap.edge' x y))                       == True
```

Comment:
I think it's best to avoid `isRight` examples -- better switch to more informative `== Right ...` instead.

Reply:
In this particular test, it would be `Right (edge x y)` in case `x < y` and `Right (edge y x)` in case `x > y`.

Reply:
Well, in fact, this particular example is incorrect, for example, `detectParts (edge 1 1) == Left [1]`. But generally, it doesn't seem we could provide a complete definition at the right side.

Reply:
Actually, `detectParts (edge 1 1) == Left [1]` should be an example.

Reply:
Aha, let's have both examples then:
```
detectParts (edge 1 1) == Left [1]
detectParts (edge 1 2) == Right (edge 1 2)
```
Very clear and useful!